### ISSUE TITLE
HTTPs in EnvironmentProxy

### ISSUE THREAD

**Issue by Rand-Random** [dotnet/runtime#44799]
I was running into an issue with this line
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L123

As the fool I am, I specified the HTTP**S**_PROXY environment variable with http**s**://[ipaddress]:[port] sadly this method will fail cause it wont remove the the https:// I specified, as it only tries to remove http://.

It will throw an `UriFormatException` (Invalid URI: The hostname could not be parsed.) on this line
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L202

because a few lines earlier a `UriBuilder` gets initialized with the `host` variable being "https://[idaddress]"
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L191

So my question is do we want to include the following check, to remove a potential https, to help fools like me?

    if (value.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
    {
        value = value.Substring(8);
    }

I would also like to question this routine
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.Windows.cs#L10

Where if `httpsProxy` it will fallback to `allProxy` would it be suitable to fallback to `httpProxy` aswell?
In my case `httpProxy` was successfully initilalized and `httpsProxy` failed to initialize and I didn't specifiy `allProxy`, and therefor I ended up with a null `httpsProxy` which caused follow up errors.

So maybe add the following line after this
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.Windows.cs#L25

`allProxy ??= httpProxy; //if allProxy is null use httpProxy as a fallBack`

This error had some annoying after effects with rather unhelpful error message about network timeout errors, and took me way to long to figure out that the letter **s** is causing the error.

**Comment by wfurt** [comment]
I'm not sure what you suggesting @Rand-Random. Using http proxy for https would be breaking change. Using https proxy is tracked by #31113.

**Comment by Rand-Random** [comment]
@wfurt 
Let me try again.

I didn't know about there being a difference in HTTPS proxy vs HTTP Proxy and just foolishly put a HTTPs URL pointing to my proxy into the environment variable.

So I thought my environment variable have to look like this
![image](https://user-images.githubusercontent.com/9868626/99508310-bc822600-2984-11eb-8537-b99b8c8baa12.png)

This will fail silently in initializing the proxy but will fail later upon requesting a `https` URL with an unhelpful network error.
I suggested ways to make it work silently instead of failing.

So, my first suggesting was to not care if the user inputs `http` or `https` and remove both.
Here only http will be removed (linked also in my original post):
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L123

My second suggestion was why not use the `httpProxy` when there wasn't a `httpsProxy` or `allProxy` specified or it failed silently, to again make it work silently.
Which was this piece of code, I provided in my original post:

`allProxy ??= httpProxy; //if allProxy is null use httpProxy as a fallBack`

(Side note: I wouldn't consider it a breaking change, at least for me, a breaking change is a change that breaks the compilation, which this change wouldn't do.)

BUT if the `EnvironmentProxy` wouldn't fail silently on my wrong input that would have also helped alot figuring out what was actually wrong, but that isn't the case.
So maybe this question is more fitting, why does `EnvironmentProxy` not throw a `NotSupportedException` when it is provided with a `https` URL?

Currently it just fails silently ending up with a `httpsProxy` being null and the following method
https://github.com/dotnet/runtime/blob/88ecbab85beead9c7d9406ee2f3ea468b1d4cd1e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L250

Returning a `null` if the scheme of the URI is `https`, so every request to a `https` URL will fail with a network error.
With 0 information that I messed up my input in the environment variables.

This is also a different behaviour as the `WebProxy` you have provided, the class `WebProxy` accepts, for what ever reason, a `https` URL and will not fail silently and hava null in the URI for the proxy Uri.

And with a `WebProxy` in use I am getting a meaning full exception upon requesting an url from here:
https://github.com/dotnet/runtime/blob/030f5933e2f82bd8fceff166fc578020943a6b11/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L398

But this line never has a chance to trigger an exception with `EnvironmentProxy` because the proxy will be null as it failed silently upon initializing.

**Comment by wfurt** [comment]
I think we should make sure there is enough tracing to diagnose issues like this. However if we blend HTTP and HTTPS together for protocol to the proxy, that would make difficult to add support for encryption to the proxy it self in  #31113 or other protocols like socks.
And some people may want to proxy HTTP and not HTTPS or vice versa. The current behavior mimics curl we used to use in 1.x and 2.x so people would get same behavior after upgrade.

**Comment by karelz** [comment]
Triage: If we did this now, we would block ourselves from future support of HTTPS proxies.

@Rand-Random if you think the exception message should be better (or tracing), please let us know. It would help us if you can tell us what it is now and what you think might be better for you to diagnose such issue easier.

Closing for now.

**Comment by Rand-Random** [comment]
@karelz
> Triage: If we did this now, we would block ourselves from future support of HTTPS proxies.

Not really, since I just suggested it as a fallback and not just do it.
So when a https URL in the https proxy evironment variable would be supported the fallback wouldn't trigger.

> @Rand-Random if you think the exception message should be better (or tracing), please let us know. It would help us if you can tell us what it is now and what you think might be better for you to diagnose such issue easier.

I don't understand your response here, because I already told you how to do it better and how it works now.
If you didn't understand what I was talking about it would help me if you could raise questions so I can help you to understand me.

----

It seems to me you and @wfurt are to focused on my idea to use the http proxy if the https failed.
Just forget about this if you don't like it, I have already pointed out 2 ways to do better exceptions/tracing.

To sum them up once again

1) In intializing the `EnvironmentProxy` throw a `NotSupportedException` if a https schema is provided
This should be a good place, as you are already checking for http: https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L123
2) Let `EnvironmentProxy` just like `WebProxy` actually accept a https URL, which will than cause an exception here
https://github.com/dotnet/runtime/blob/88ecbab85beead9c7d9406ee2f3ea468b1d4cd1e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L398

To do this those lines need to be adjusted
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L123
https://github.com/dotnet/runtime/blob/69d61271f900afd7ae6dea7f0c5e57d43090f9dd/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs#L191

**Comment by karelz** [comment]
> If you didn't understand what I was talking about it would help me if you could raise questions so I can help you to understand me.

I think the key source of misunderstanding here is that we are mixing scenario and behavior discussion with solution discussions. So, let's step back and let's clarify scenarios and current behavior first:

I can see these 3 main relevant scenarios:
1. User wants to use HTTP**S** proxy (which is not supported by .NET 5, but will likely be supported in future)
2. User accidentally puts HTTP**S** prefix in front of proxy URL (which is just HTTP) -- the proxy server won't respond on that port or won't work
3. User uses invalid proxy URL (the URL does not exist)

Current behavior:
1. .NET 5 will ignore the proxy setting and create connections directly, without using any proxy. As a result, the connections will likely fail (assuming the network is protected, making the proxy the only way to make outgoing calls). User has to troubleshoot (how?) to discover that HTTP**S** proxies are not supported.
    - Note: Given that HTTP**S** proxies are rare, this is fairly rare scenario at the moment, but might be more common with in future, especially given .NET 6 is LTS and will be supported for a few years.
    - Assumption: This is how curl works -- @wfurt is it correct statement?
2. Behavior same as in above scenario [1] -- proxy setting will be ignored, connections will likely fail. User has to troubleshoot (how?) to discover they made a mistake.
    - Note: This is fairly rare scenario as we got 1 user complaint so far (since .NET Core 3.0 release where we started supporting environment variables). Albeit, it is rather nasty.
3. User gets exception `System.Net.Http.HttpRequestException: No such host is known. (specific_wrong_proxy_url:80)`

In some future version of .NET, HTTP**S** proxies will be supported (work tracked by #31113). Here's expected impact on the 3 scenarios above:
1. This scenario will just start working. No need to troubleshoot anything.
2. This scenario will start failing in a similar way as scenario [3] does today (see above), because the proxy won't listen on the correct port 8080.
3. Same error as does today.

What would be the ideal current behavior?
I don't think we can distinguish easily between scenario [1] and [2]. If we decide to change behavior, it will apply to both scenarios.
The proposal is to throw exception (e.g. `UnsupportedException`) when HTTP**S** proxy is encountered and not supported. On one hand it sounds reasonable, on the other hand it seems to be breaking change against previous curl implementation.

The key question is: What do users expect? Is the assumed curl (and our current) behavior somehow valuable, or are we lacking just more reasonable troubleshooting path? (aka exception) ... Personally I think throwing exception might be reasonable, unless @wfurt knows about a scenario / user expectation I missed.

**Comment by Rand-Random** [comment]
@karelz
Sorry for my late response.

> Albeit, it is rather nasty.
Have to agree with this 100%, took me way to long to figure out what went wrong on my side.
Though of course it was my error that I just put a https url into the https_proxy environment variable without thinking about it.

> On one hand it sounds reasonable, on the other hand it seems to be breaking change against previous curl implementation.
As you said your self if someone has the wrong input like I did, the current behaviour will only make issues that are hard to track the root cause, so throwing an exception IMHO would introduce an issue earlier that is easier to detect and fix.
Additionally, I am the only one who fell into this trap, so there shouldn't be customers out there that would be affected by this breaking change.

> curl
You and @wfurt have mentioned `curl` and that it is the same behaviour as the `EnvironmentProxy`.
I would like to raise the question why is it important or desirable to have the same behaviour in the `EnvironmentProxy` but not in the `WebProxy`?
As I already mentioned the `WebProxy` just simply accepts an input of https without any validation, and lets the `HttpConnectionPoolManager.cs` throw a meaningful exception, as can be seen in the issue you both mentioned for tracking https support #31113.

**Comment by wfurt** [comment]
curl was used to power HttpClient on Unix in earlier .NET versions. So we wanted to carry over the behavior for compat reasons e.g. the variables would work the same way after upgrade. It also seems like many other packages adopted this on Unix so it feels like it would make sense to be consistent in that environment.
